src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -2443,105 +2442,41 @@ internal void EnsureAnonymousTypeTemplates(CancellationToken cancellationToken)
-                    // stream that this method was called with.
-                    Stream retStream;
-                    if (!metadataOnly && IsRealSigned && this.Options.StrongNameProvider.Capability == SigningCapability.SignsStream)
-                    {
-                        Debug.Assert(Options.StrongNameProvider != null);
-
-                        try
-                        {
-                            signingInputStream = Options.StrongNameProvider.CreateInputStream();
-                        }
-                        catch (IOException e)
-                        {
-                            throw new Cci.PeWritingException(e);
-                        }
-
-                        retStream = signingInputStream;
-                    }
-                    else
-                    {
-                        signingInputStream = null;
-                        retStream = peStream;
-                    }
-
-                    return retStream;
+                    Stream ret;
+                    (peStream, signingInputStream, ret) = GetPeStream(metadataDiagnostics, peStreamProvider, metadataOnly);
+                    return ret;
                 };
 
-                Func<Stream> getRefPeStream;
-                if (metadataPEStreamProvider != null)
+                Func<Stream> getRefPeStream = () =>
                 {
-                    getRefPeStream = () =>
-                    {
-                        if (metadataDiagnostics.HasAnyErrors())
-                        {
-                            return null;
-                        }
+                    return GetAuxPeStream(metadataDiagnostics, metadataPEStreamProvider);
+                };
 
-                        refPeStream = metadataPEStreamProvider.GetOrCreateStream(metadataDiagnostics);
-                        Debug.Assert(refPeStream != null || metadataDiagnostics.HasAnyErrors());
-                        return refPeStream;
-                    };
-                }
-                else
+                var includePortablePdbStream = moduleBeingBuilt.DebugInformationFormat == DebugInformationFormat.PortablePdb && pdbStreamProvider != null;
+                Func<Stream> getPortablePdbStream = () =>
                 {
-                    getRefPeStream = null;
-                }
+                    portablePdbStream = GetAuxPeStream(metadataDiagnostics, pdbStreamProvider);
+                    return portablePdbStream;
+                };
 
                 try
                 {
                     if (SerializePeToStream(
                         moduleBeingBuilt,
                         metadataDiagnostics,
-                        this.MessageProvider,
+                        MessageProvider,
                         getPeStream,
-                        getRefPeStream,
-                        getPortablePdbStream,
+                        metadataPEStreamProvider != null ? getRefPeStream : null,
```

Comment:
I like this check as a ternary because it seems simple, but the second check seems like we're delaying a calculation away from its original intent for too long.

Reply:
Fixed.

Reply:
Done.